src/main/java/com/timgroup/statsd/UnixStreamClientChannel.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Override
+    public synchronized int write(ByteBuffer src) throws IOException {
+        connectIfNeeded();
+
+        int size = src.remaining();
+        if (size == 0) {
+            return 0;
+        }
+        delimiterBuffer.clear();
+        delimiterBuffer.putInt(size);
+        delimiterBuffer.flip();
+
+        try {
+            long deadline = System.nanoTime() + timeout * 1_000_000L;
+            if (writeAll(delimiterBuffer, true, deadline) > 0) {
+                writeAll(src, false, deadline);
+            }
+        } catch (IOException e) {
+            disconnect();
+            throw e;
+        }
+
+        return size;
+    }
+
+    /**
+     * Writes all bytes from the given buffer to the channel.
+     * @param bb buffer to write
+     * @param canReturnOnTimeout if true, we return if the channel is blocking and we haven't written anything yet
+     * @param deadline deadline for the write
+     * @return number of bytes written
+     * @throws IOException if the channel is closed or an error occurs
+     */
+    public int writeAll(ByteBuffer bb, boolean canReturnOnTimeout, long deadline) throws IOException {
+        int remaining = bb.remaining();
+        int written = 0;
+        while (remaining > 0) {
+            int read = delegate.write(bb);
+
+            // If we haven't written anything yet, we can still return
+            if (read == 0 && canReturnOnTimeout && written == 0) {
+                return written;
+            }
+
+            remaining -= read;
+            written += read;
+
+            if (deadline > 0 && System.nanoTime() > deadline) {
+                throw new IOException("Write timed out");
+            }
+        }
+        return written;
+    }
+
+    private void connectIfNeeded() throws IOException {
+        if (delegate == null) {
+            connect();
+        }
+    }
+
+    private void disconnect() throws IOException {
+        if (delegate != null) {
+            delegate.close();
+            delegate = null;
+        }
+    }
+
+    private void connect() throws IOException {
+        if (this.delegate != null) {
+            try {
+                disconnect();
+            } catch (IOException e) {
+                // ignore to be sure we don't stay with a broken delegate forever.
+            }
+        }
+
+        UnixSocketChannel delegate = UnixSocketChannel.create();
+
+        long deadline = System.currentTimeMillis() + connectionTimeout;
+        if (connectionTimeout > 0) {
+            // Set connect timeout, this should work at least on linux
+            // https://elixir.bootlin.com/linux/v5.7.4/source/net/unix/af_unix.c#L1696
+            // We'd have better timeout support if we used Java 16's native Unix domain socket support (JEP 380)
+            delegate.setOption(UnixSocketOptions.SO_SNDTIMEO, connectionTimeout);
+        }
+        delegate.connect(address);
+        while (!delegate.finishConnect()) {
```

Comment:
According the code, connect could be returning false:
```
        if (Native.connect(getFD(), remote, remote.length()) != 0) {
            Errno error = Errno.valueOf(LastError.getLastError(jnr.ffi.Runtime
                    .getSystemRuntime()));

            switch (error) {
            case EAGAIN:
            case EWOULDBLOCK:
                return false;
```

Although I'm not certain on which systems this would happen. So this seems worth handling.

This would all be better if we could use Java's native UDS support

Reply:
> Although I'm not certain on which systems this would happen.

This happens on Linux when the connection times out. Bsd and mac don't block on connect  at all (for unix sockets) and return ECONNREFUSED instead when the listen queue is full, which would be an exception.

Reply:
Do you think we can remove this loop and check return value from connect?

Reply:
If `connect()` returns `false` (and not an exception) we're supposed to try again using `finishConnect()` no ? We could skip the loop if `connect()` returns `true`

I think it could return `false` without hitting the timeout if the queue is full and the socket was set to non-blocking mode (and it's unclear that there's a guarantee it won't be)

Reply:
> If connect() returns false (and not an exception) we're supposed to try again using finishConnect() no ? We could skip the loop if connect() returns true

Techincally, yes. If so, we should check for timeout before calling finishConnect(), to avoid waiting twice as long, because both connect and finishConnect block. But for the specific situation we have here (blocking unix socket), I think it is redundant, as the first call would wait long enough.

> I think it could return false without hitting the timeout if the queue is full and the socket was set to non-blocking mode (and it's unclear that there's a guarantee it won't be)

We don't set non-blocking mode ourselves, and we can very much count on the fact that it won't be set by default.

Reply:
Changed to what you suggest